### PR TITLE
CA-285270: Enable iso file download in install update wizard

### PR DESCRIPTION
**Pull request by jijiang**
Signed-off-by: Ji Jiang <ji.jiang@citrix.com>

### CODE REVIEW

File: XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs [hunk 1]
```diff
@@ -34,6 +34,7 @@
 using System.ComponentModel;
 using System.Threading;
 using System.IO;
+using System.Web.UI.WebControls;
```
Comment (michael2012z) [inline XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs]: It was not used, should be removed.

File: XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs [hunk 2]
```diff
@@ -68,7 +70,9 @@ public DownloadAndUnzipXenServerPatchAction(string patchName, Uri uri, string ou
             updateName = patchName;
             address = uri;
             downloadUpdate = address != null;
-            zippedFileName = outputFileName;
+            nonExtractUpdate = downloadUpdate &&
+                               Array.Exists(updateFileExtensions, item => address.ToString().Contains(item));
```
Comment (michael2012z) [inline XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs]: If an URL contains an extension, it must be a file that doesn't need to be extracted. Am I right?
I think we'd better guarantee the URL contains "iso" (for example) as an extension (at the end). If the URL is like "xxxxxisoxxxx", it also pass the check but it was not what we want.
Please correct me if my understanding is incorrect.

File: XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs [hunk 3]
```diff
@@ -217,9 +221,7 @@ protected override void Run()
         {
             if (downloadUpdate)
             {
-                log.InfoFormat("Downloading update '{0}' (from from '{1}') to '{2}'", updateName, address, zippedFileName);
-
-
+                log.InfoFormat("Downloading update '{0}' (from from '{1}') to '{2}'", updateName, address, outputFileName);
```
Comment (kc284) [inline XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs]: Double `from`.

File: XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs (hunk 2 above)
Comment (kc284) [inline XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs]: Or simply `updateFileExtensions.Any(item => address.ToString().Contains(item))`

File: XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs [hunk 4]
```diff
@@ -232,11 +234,23 @@ protected override void Run()
...
-            log.DebugFormat("Extracting XenServer patch '{0}'", updateName);
-            Description = string.Format(Messages.DOWNLOAD_AND_EXTRACT_ACTION_EXTRACTING_DESC, updateName);
-            ExtractFile();
-            log.DebugFormat("Extracting XenServer patch '{0}' completed", updateName);
-
+            if (nonExtractUpdate)
+            {
+                string newFilePath = Path.Combine(Path.GetDirectoryName(outputFileName), updateName);
+                if (File.Exists(newFilePath))
+                    File.Delete(newFilePath);
```
Comment (kc284) [inline XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs]: Can we include some exception handling for these IO operations (Delete and Move).

File: XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs [hunk 5]
```diff
@@ -49,10 +50,11 @@ public class DownloadAndUnzipXenServerPatchAction : AsyncAction
         private Random random = new Random();
 
         private readonly Uri address;
-        private readonly string zippedFileName;
+        private readonly string outputFileName;
         private readonly string updateName;
         private readonly string[] updateFileExtensions;
         private readonly bool downloadUpdate;
+        private readonly bool nonExtractUpdate;
```
Comment (kc284) [inline XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs]: Language improvement: let's rename this to `doNotExtractUpdate`.

File: XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs [hunk 6]
```diff
@@ -203,23 +220,23 @@ private void ExtractFile()
                     iterator.Dispose();
 
                 if (downloadUpdate)
-                    File.Delete(zippedFileName);
+                    File.Delete(outputFileName);
```
Comment (kc284) [inline XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs]: Apologies, I only saw this now: can we wrap this in a try-catch block too? An empty catch block will suffice:
```
if (downloadUpdate)
{
try { File.Delete(outputFileName); }
catch { }
}
```

**Comment by jijiang** [comment]
Tested. Works fine now.

**Pull request merged**